pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+
+- Union types (represented by ``A | B`` according to PEP 604) are
+  valid in function signature returns, so we need to allow operators
+  in the return position for consistency.
+- Given that operators bind more tightly than ``->`` it is correct
+  that a type like ```bool | () -> bool`` must be a syntax error. We
+  should be sure the error message is clear because this may be a
+  common mistake.
+- Associating ``->`` to the right, rather than requiring explicit
+  parentheses, is consistent with other languages like TypeScript and
+  respects the principle that valid expressions should normally be
+  substitutable when possible.
+
+``async`` Keyword
+~~~~~~~~~~~~~~~~~
+
+All of the binding rules still work for async callable types::
+
+    (int) -> async (float) -> str | bool
+    (int) -> (async (float) -> (str | bool))
+
+    def f() -> async (int, str) -> bool: pass
+    def f() -> (async (int, str) -> bool): pass
+
+    def f() -> async (int) -> async (str) -> bool: pass
+    def f() -> (async (int) -> (async (str) -> bool)): pass
+
+
+Trailing Commas
+~~~~~~~~~~~~~~~
+
+- Following the precedent of function signatures, putting a comma in
+  an empty arguments list is illegal, ``(,) -> bool`` is a syntax
+  error.
+- Again following precedent, trailing commas are otherwise always
+  permitted::
+
+
+    ((int,) -> bool == (int) -> bool
+    ((int, **P,) -> bool == (int, **P) -> bool
+    ((...,) -> bool) == ((...) -> bool)
+
+Allowing trailing commas also gives autoformatters more flexibility
+when splitting callable types across lines, which is always legal
+following standard python whitespace rules.
+
+
+Disallowing ``...`` as an Argument Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under normal circumstances, any valid expression is permitted where we
+want a type annotation and ``...`` is a valid expression. This is
+never semantically valid and all type checkers would reject it, but
+the grammar would allow it if we did not explicitly prevent this.
+
+We decided that there were compelling reasons to prevent it: - The
+semantics of ``(...) -> bool`` are different from ``(T) -> bool`` for
+any valid type T: ``(...)`` is a special form indicating
+``AnyArguments`` whereas ``T`` is a type parameter in the arguments
+list.  - ``...`` is used as a placeholder default value to indicate an
+optional argument in stubs and Callback Protocols. Allowing it in the
+position of a type could easily lead to confusion and possibly bugs
+due to typos.
+
+Since ``...`` is meaningless as a type and there are usability
+concerns, our grammar rules it out and the following is a syntax
+error::
+
+    (int, ...) -> bool
+
+Incompatibility with other possible uses of ``*`` and ``**``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any
+future proposal using a bare ``**<some_type>`` to type
+``kwargs``. This seems acceptable because:
```

Comment:
Hm. Occasionally we hear from someone who has a TypedDict that they want to use as the type for some function's kwargs. If we wanted to support that for a function *type*'s kwargs we'd be hosed -- the obvious notation would be `(**TD) -> bool`. Or maybe that would just work, and what you're ruling out here is use of `(**int) -> bool` to type e.g. `def f(**args: int) -> bool: ...`?

Reply:
Hmm... Technically, there is nothing to prevent us from supporting `**KwargsTypedDict` (or even `**int`) in the future. Type checkers could distinguish `**KwargsTypedDict` and `**P` based on the actual values of the variables. 

The main reason we have `**P` instead of a bare `P` is to make it syntactically obvious that it is a ParamSpec (like `Callable[P, R]`). Otherwise, users might have trouble realizing that P is a ParamSpec in `(Foo, P) -> int`.

We could update the PEP (in a new PR) with the above.

Reply:
I'm on the fence about this. Technically, there's nothing prevent us from using `**<some_type>` with divergent semantics depending on the information contained within `<some_type>`, type-checkers could handle this.

But it would be pretty unusual for python to let the semantics of the same syntax vary that much - a `ParamSpec` is a pretty different concept from typing the `kwarg` variadic argument.

My guess is that, in an extended syntax, we would work around this by requiring names for the vararg and kwarg,
so that we would use:
```
(**kwags: int) -> bool
(**kwargs: **MyTypedDict) -> bool
```

I think we should update the PEP to clarify that it's technically still possible to use `**int` and `**MyTypedDict` but I don't want to imply this is a good idea or say anything that's likely to prompt much discussion of it.

Reply:
For what it's worth, I don't see the issue with allowing `**MyTypedDict` in the future: just like `**P`, it means "this is the type for all of the kwargs". Perhaps to keep this option open we should avoid using the name "paramspec" in the AST and runtime objects.

Reply:
> But it would be pretty unusual for python to let the semantics of the same syntax vary that much

Would it, though? We have `1+2 == 3`, vs `"a" + "b" == "ab"`. We have `a[i]` (indexing list or tuple) vs. `d[key]` (lookup in dict or mapping) vs. `list[int]` (generic type). Etc. So I think there are enough precedents -- we don't need to worry about this looking odd.